### PR TITLE
Support GC.at_preserve

### PR DESCRIPTION
**Pull request by vchuravy**

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -10,6 +10,7 @@ Enzyme_jll = "7cc45869-7501-5eee-bdea-0790c847d4ef"
 GPUCompiler = "61eb1bfa-7361-4325-ad38-22787b887f55"
 LLVM = "929cbde3-209d-540e-8aea-75f648917ca0"
 Libdl = "8f399da3-3557-5675-b5ff-fb832c97cbdb"
+MPI = "da04e1cc-30fd-572f-bb4f-1f8673147195"
```
Comment (vchuravy) [inline Project.toml:13]:
```suggestion
```

**Comment by wsmoses** [comment]
@vchuravy rebased, this now passes

File: src/compiler.jl [hunk 2]
```diff
@@ -1720,6 +1939,8 @@ function GPUCompiler.codegen(output::Symbol, job::CompilerJob{<:EnzymeTarget};
         augmented_primalf = nothing
     end
 
+    # fix_inlining!(mod)
```
Comment (vchuravy) [inline src/compiler.jl:1942]: So we can remove `fix_inling` from this PR?

File: src/compiler/validation.jl [hunk 3]
```diff
@@ -320,6 +320,33 @@ function check_ir!(job, errors, imported, inst::LLVM.CallInst, known_fns)
...
+                if ptr == cglobal(:jl_array_del_beg)
+                    fn = "jl_array_del_beg"
+                end
+                if ptr == cglobal(:jl_array_del_end)
+                    fn = "jl_array_del_end"
+                end
+                if ptr == cglobal(:jl_array_del_at)
+                    fn = "jl_array_del_at"
+                end
+		if ptr == cglobal(:jl_value_ptr)
```
Comment (vchuravy) [inline src/compiler/validation.jl:341]: Formatting

File: src/compiler.jl [hunk 4]
```diff
@@ -975,28 +1101,121 @@ function annotate!(mod)
...
+
     for rfn in ("jl_object_id_", "jl_object_id")
         if haskey(fns, rfn)
             fn = fns[rfn]
             push!(function_attributes(fn), LLVM.EnumAttribute("readonly", 0; ctx))
         end
     end
+
+    # MPI_Irecv
+    for rfn in ("MPI_Irecv", "PMPI_Irecv")
```
Comment (vchuravy) [inline src/compiler.jl:1142]: Can we split those out?

**Review by vchuravy — COMMENTED** [review]
I am going to split the MPI changes into their own PR

File: src/compiler.jl (hunk 2 above)
Comment (wsmoses) [reply src/compiler.jl:1942]: it was to attempt to fix mpi misbehaving, but I don't think it ended up being that issue. Regardless fine to split with MPI

**Pull request merged**